### DISCUSSION TITLE
Freezing the templates folder into the binary

### DISCUSSION THREAD

**Discussion by temphstarmans** [miguelgrinberg/microdot#227] [Q&A]
I am thinking of freezing the templates directory used by utemplate into the micropython binary, so I can roll updates via ```github:glenn20/micropython-esp32-ota/mip/ota/mpy```. Do you think this is a good idea? Typically, all my static resources are small.   There are ways of freezing non python code into the binary, see https://github.com/orgs/micropython/discussions/10707 . Especially, https://github.com/bixb922/freezefs seems interesting as it also calculates ram usage.  I think it requires small changes to microdot/utemplate.py, as I need to pick up the template from the memory and compile to the flash memory.

Another option I see is to roll the OTA update and then install the non-python assets via a mip install and package.json https://docs.micropython.org/en/latest/reference/packages.html

**Comment by miguelgrinberg** [comment]
I have good news for you.

Templates need to be compiled before they are used. The compiled code is actually Python. You can compile your templates on your desktop/laptop if you want, then configure the uTemplate integration in Microdot to not attempt to compile, just load the pre-compiled versions that are Python code. I explained how to do this [in this comment](https://github.com/miguelgrinberg/microdot/discussions/198#discussioncomment-8093663).

Since this all ends up being Python code, it's absolutely no issue to freeze your templates.

**Comment by temphstarmans** [comment]
Use 
```
python -m freezefs  templates frozen_templates.py --on-import=extract --compress
```
In boot.py you can then do uncompress, which will automatically create the directory templates.
```
import frozen_templates
```
I disagree with your solution because I believe that the resulting python compile result depends on the variables you pass.
Let me know if this is incorrect.

**Comment by miguelgrinberg** [comment]
I don't know what freezefs is, I have never used it. By "freezing" I meant compiling a custom firmware as explained here: https://docs.micropython.org/en/latest/reference/manifest.html.

**Comment by temphstarmans** [comment]
It's called freezefs, because it allows you to freeze a non-python file or directory into your micropython image.
So freezefs converts the folder to a python file. This python file can then be frozen into the micropython image.

My current procedure is. I call manifest.py, this freezes the required folders and pushes these into my package.
I then build the image. I will add my manifest.py.. still perfecting it. On closer inspection, it seems not needed for the templates folder. perhaps only the staic folder.

**Comment by miguelgrinberg** [comment]
uTemplate compiles templates to standard Python files, as I said above. No need to freeze the actual template files, you can pre-compile and then freeze the compiled Python files. That also eliminates the need for uTemplate to compile the templates in your device.

**Reply by ghost** [reply]
So how do I pre-compile the template files in advance?

**Reply by miguelgrinberg** [reply]
You can just use the templates normally and they get compiled on first use. Alternatively, uTemplate has a command to compile templates, see utemplate_util.py in https://github.com/pfalcon/utemplate.

**Reply by hstarmans** [reply]
@EvansMike 
I finished a webserver which freezes parts of the templates and images into the micocropython binary. I showcase my microdot webserver here https://www.youtube.com/watch?v=p6ITKJ2BUik
The build procedure is outline in the manifest py, see https://github.com/hstarmans/esp32_hexastorm/blob/master/src/manifest.py ..

**Reply by ghost** [reply]
@hstarmans  Thanks for that. It looks useful for freeing up even more memory.
